databricks/koalas/mlflow.py

```diff
@@ -0,0 +1,166 @@
+#
+# Copyright (C) 2019 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+MLflow-related functions to load models and apply them to Koalas dataframes.
+"""
+from mlflow.pyfunc import PythonModel, load_pyfunc, spark_udf
+from pyspark.sql.types import DataType
+import pandas as pd
+import numpy as np
+
+from databricks.koalas.utils import lazy_property, default_session
+from databricks.koalas import Series, DataFrame
+from databricks.koalas.typedef import as_spark_type
+
+__all__ = ["PythonModelWrapper", "load_model"]
+
+
+class PythonModelWrapper(object):
+    """
+    A wrapper around MLflow's Python object model.
+
+    This wrapper acts as a predictor on koalas
+
+    """
+    def __init__(self, path, run_id, return_type_hint):
+        self._path = path  # type: str
+        self._run_id = run_id  # type: str
+        self._return_type_hint = return_type_hint
+
+    @lazy_property
+    def _return_type(self) -> DataType:
+        hint = self._return_type_hint
+        # The logic is simple for now, because it corresponds to the default
+        # case: continuous predictions
+        # TODO: do something smarter, for example when there is a sklearn.Classifier (it should
+        # return an integer or a categorical)
+        # We can do the same for pytorch/tensorflow/keras models by looking at the output types.
+        # However, this is probably better done in mlflow than here.
+        if hint == 'infer' or not hint:
+            hint = np.float64
+        return as_spark_type(hint)
+
+    @lazy_property
+    def _model(self) -> PythonModel:
+        return load_pyfunc(self._path, self._run_id)
+
+    @lazy_property
+    def _model_udf(self):
+        spark = default_session()
+        return spark_udf(spark, self._path, self._run_id, result_type=self._return_type)
+
+    def __str__(self):
+        return "PythonModelWrapper({})".format(str(self._model))
+
+    def __repr__(self):
+        return "PythonModelWrapper({})".format(repr(self._model))
+
+    def predict(self, data):
+        """
+        Returns a prediction on the data.
+        If the data is a pandas Dataframe, the return is a pandas series (TODO: np.array?)
```

Comment:
The return value is pandas.Series.

Reply:
Unfortunately, in the case of pandas dataframe input, the output can be whatever (and the doctest shows that sklearn returns a np.array instead).